examples/cpp/pyperf/PyPerfBPFProgram.cc

```diff
@@ -313,6 +317,32 @@ on_event(struct pt_regs* ctx) {
   event->stack_status = STACK_STATUS_ERROR;
   event->error_code = ERROR_NONE;
 
+  struct task_struct const *const task = (struct task_struct *)bpf_get_current_task();
+
+  // Get raw native user stack
+  struct pt_regs user_regs = *ctx;
+  if (!user_mode(&user_regs)) {
+    // The third argument is equivalent to `task_pt_regs(task)`
+    bpf_probe_read_kernel(
+        &user_regs, sizeof(user_regs),
+        (struct pt_regs *)((unsigned long)(task->stack) + THREAD_SIZE -
+                           TOP_OF_KERNEL_STACK_PADDING) - 1);
+  }
+
+  event->user_sp = user_regs.sp;
+  event->user_ip = user_regs.ip;
+
+  // Copy one page at the time - if one fails we don't want to lose the others
+  int i;
+  for (i = 0; i < __USER_STACKS_PAGES__; ++i) {
+    if (bpf_probe_read_user(
+            event->raw_user_stack + i * PAGE_SIZE, PAGE_SIZE,
+            (void *)((user_regs.sp & PAGE_MASK) + (i * PAGE_SIZE))) < 0) {
```

Comment:
Masking out `& PAGE_SIZE` means we'll usually get lots of the "unused" stack. That's fine - but what about the other case? If `sp` is e.g at `0x1000`, masking out doesn't bring it any lower; no chance libunwind might require stuff from `0x998` for instance? https://en.wikipedia.org/wiki/Red_zone_(computing)

Perhaps we should get the stack from a constant offset of `sp`?